In [1]:
import importlib
import os
import pickle
import random

import numpy as np
import pandas as pd
import torch
import wandb
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from torch.utils.data import DataLoader
from torchvision.transforms import (
    Compose,
    RandomAffine,
    RandomHorizontalFlip,
    RandomResizedCrop,
    RandomRotation,
    RandomVerticalFlip,
)

from src import model, modules, utils

importlib.reload(model)
importlib.reload(modules)
importlib.reload(utils)

# pylint: disable = C0413
from src.model import MammoCNN
from src.utils import (
    BIRADS_MIN,
    LOSS_WEIGHTS,
    MammoCNNDataset,
    calibrate_loss_scales,
    effective_contributions,
    evaluate,
    get_coral_cutpoint_weights,
    get_findings_weights,
    get_lds_weights,
    train,
)

CLINICAL_FINDINGS = [
    "Mass",
    "Global Asymmetry",
    "Architectural Distortion",
    "Nipple Retraction",
    "Suspicious Calcification",
    "Focal Asymmetry",
    "Asymmetry",
    "Skin Thickening",
    "Suspicious Lymph Node",
    "Skin Retraction",
]

In [2]:
SEED: int = 42
BATCH_SIZE: int = 48
NUM_WORKERS: int = 8
DEVICE: torch.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
IMAGE_SIZE: list[int] = [1280, 1024]  # (H, W)


# Set seed to ensure consistency across runs
def seed_worker(worker_id):
    worker_seed = SEED + worker_id
    np.random.seed(worker_seed)
    random.seed(worker_seed)


g = torch.Generator()
g.manual_seed(SEED)
random.seed(SEED)
np.random.seed(SEED)

torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed(SEED)
    torch.cuda.manual_seed_all(SEED)

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

#### Split the dataset by using 600 patients for a hold out test set. The remaining 4,400 patients are randomly partitioned into a MAMMO CNN training set, a Classifier training set, a Triage training set, and a validation set of 60%, 15%, 15%, and 10%, respectively.

In [3]:
metadata_path = os.path.join("vindr-mammo", "breast_metadata.parquet")
metadata_df = pd.read_parquet(metadata_path)

# Create a patient level dataframe
patient_df = (
    metadata_df[["study_id", "patient_cancer", "Patient's Age"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

# Hold out test set 12%
train_val, test = train_test_split(
    patient_df,
    test_size=0.12,
    random_state=SEED,
    stratify=patient_df["patient_cancer"],
)

# Of the remaining 88%, split: 60% | 15% | 15% | 10%
# Sequential fractions:
#   val:             10/100 = 0.10  of remaining
#   triage_train:    15/90  ≈ 0.1667 of what's left after val
#   classifier_train:15/75  = 0.20  of what's left after val + triage
#   cnn_train:       60/60  = 1.0   (everything left)

# Split the remaining 4,400 patients
remaining_1, val = train_test_split(
    train_val,
    test_size=0.10,
    random_state=SEED,
    stratify=train_val["patient_cancer"],
)

remaining_2, triage_train = train_test_split(
    remaining_1,
    test_size=round(15 / 90, 6),  # ≈ 0.1667
    random_state=SEED,
    stratify=remaining_1["patient_cancer"],
)

cnn_train, classifier_train = train_test_split(
    remaining_2,
    test_size=round(15 / 75, 6),  # = 0.20
    random_state=SEED,
    stratify=remaining_2["patient_cancer"],
)

cnn_train.reset_index(drop=True, inplace=True)
classifier_train.reset_index(drop=True, inplace=True)
triage_train.reset_index(drop=True, inplace=True)
val.reset_index(drop=True, inplace=True)
test.reset_index(drop=True, inplace=True)

print(f"Number of patients in CNN training set: {len(cnn_train)}")
print(f"Number of patients in Classifier training set: {len(classifier_train)}")
print(f"Number of patients in Triage training set: {len(triage_train)}")
print(f"Number of patients in Validation set: {len(val)}")
print(f"Number of patients in Test set: {len(test)}")

Number of patients in CNN training set: 2639
Number of patients in Classifier training set: 660
Number of patients in Triage training set: 661
Number of patients in Validation set: 440
Number of patients in Test set: 600


#### Map the patient level splits back to the breast level dataframe

In [4]:
cnn_patients = cnn_train["study_id"].tolist()
classifier_patients = classifier_train["study_id"].tolist()
triage_patients = triage_train["study_id"].tolist()
val_patients = val["study_id"].tolist()
test_patients = test["study_id"].tolist()

# Map the patient level labels back to the original metadata
cnn_df = metadata_df[metadata_df["study_id"].isin(cnn_patients)].reset_index(drop=True)
classifier_df = metadata_df[
    metadata_df["study_id"].isin(classifier_patients)
].reset_index(drop=True)
triage_df = metadata_df[metadata_df["study_id"].isin(triage_patients)].reset_index(
    drop=True
)
val_df = metadata_df[metadata_df["study_id"].isin(val_patients)].reset_index(drop=True)
test_df = metadata_df[metadata_df["study_id"].isin(test_patients)].reset_index(
    drop=True
)

# Merge all training dataframes
train_df = pd.concat([cnn_df, classifier_df, triage_df], ignore_index=True)

# Standardize the age column
scaler = StandardScaler()
scaler.fit(train_df[["Patient's Age"]])
cnn_df["Patient's Age"] = scaler.transform(cnn_df[["Patient's Age"]])
classifier_df["Patient's Age"] = scaler.transform(classifier_df[["Patient's Age"]])
triage_df["Patient's Age"] = scaler.transform(triage_df[["Patient's Age"]])
val_df["Patient's Age"] = scaler.transform(val_df[["Patient's Age"]])
test_df["Patient's Age"] = scaler.transform(test_df[["Patient's Age"]])

# Save the age scaler
age_encoder_path = os.path.join("models", "age_encoder.pkl")

with open(age_encoder_path, "wb") as f:
    pickle.dump(scaler, f)

# Keep the std so age MAE can be reported in years, as the paper does (5.97 yrs)
AGE_STD: float = float(scaler.scale_[0])

cnn_df.rename(columns={"Patient's Age": "age"}, inplace=True)
classifier_df.rename(columns={"Patient's Age": "age"}, inplace=True)
triage_df.rename(columns={"Patient's Age": "age"}, inplace=True)
val_df.rename(columns={"Patient's Age": "age"}, inplace=True)
test_df.rename(columns={"Patient's Age": "age"}, inplace=True)

print(f"age std (years): {AGE_STD:.2f}")
print(f"BI-RADS range: {cnn_df['breast_birads'].min()}-{cnn_df['breast_birads'].max()}")
print(f"density range: {cnn_df['breast_density'].min()}-{cnn_df['breast_density'].max()}")

age std (years): 9.09
BI-RADS range: 1-5
density range: 1-4


In [5]:
# 0-based ordinal levels for the weight helpers (the Dataset does this shift
# internally; here it is needed to count per-cutpoint positives)
cnn_suspicion_levels = cnn_df["breast_birads"].to_numpy() - BIRADS_MIN

# Per-cutpoint pos_weight for the ordinal (CORAL) suspicion head
suspicion_weights = get_coral_cutpoint_weights(cnn_suspicion_levels, n_cutpoints=4)

# Per-class weights for the multilabel findings head, clipped so the 17
# Skin Retraction positives cannot dominate the head
findings_matrix = np.stack(cnn_df["finding_vector"].to_numpy())
findings_weights = get_findings_weights(findings_matrix)

# LDS weights for density. kernel_size=3 because density has only 4 bins
density_lds = get_lds_weights(
    (cnn_df["breast_density"].to_numpy() - 1).tolist(),
    label_min=0,
    label_max=3,
    kernel_size=3,
    reweight="sqrt_inverse",
)
density_weights = torch.tensor(
    [density_lds[level] for level in range(4)], dtype=torch.float32
)

# MONAI applies alpha as: target * alpha + (1 - target) * (1 - alpha)
# so ALPHA is the weight on the POSITIVE (malignant) class and 1 - ALPHA the
# weight on the negative one. Two defensible readings:
#
#   ALPHA = POSITIVE_RATE -> positives 0.05, negatives 0.95  (pos:neg 0.05)
#       follows Lin et al., who found alpha < 0.5 best for the rare class at
#       gamma=2 because gamma already suppresses easy negatives hard.
#   ALPHA = 1 - POSITIVE_RATE  -> positives 0.95, negatives 0.05  (pos:neg 19.2)
#       inverse-frequency weighting, which is how the MAMMO paper describes
#       its alpha ("the inverse class frequency tuning parameter", Eq. 8).
#
# These differ by ~370x in the relative weight on malignancy, so it is worth
# running both and comparing diagnosis AP. Flip the line below to switch.
POSITIVE_RATE: float = float(cnn_df["diagnosis"].mean())
ALPHA: float = POSITIVE_RATE
# ALPHA: float = 1.0 - POSITIVE_RATE
GAMMA: float = 2.0

print("suspicion cutpoint pos_weights: ", [round(w, 2) for w in suspicion_weights.tolist()])
print("findings class weights: ", [round(w, 2) for w in findings_weights.tolist()])
print("density class weights: ", [round(w, 2) for w in density_weights.tolist()])
print(
    f"\nfocal alpha {ALPHA:.4f} -> malignant weight {ALPHA:.4f}, "
    f"benign weight {1 - ALPHA:.4f} (pos:neg = {ALPHA / (1 - ALPHA):.4f})"
)

loss_kwargs = {
    "loss_weights": LOSS_WEIGHTS,
    "suspicion_weights": suspicion_weights,
    "findings_weights": findings_weights,
    "density_weights": density_weights,
    "alpha": ALPHA,
    "gamma": GAMMA,
}

# Data transforms
ASPECT = IMAGE_SIZE[1] / IMAGE_SIZE[0]  # W / H
augmentations = Compose(
    [
        # Image flips
        RandomHorizontalFlip(p=0.5),
        RandomVerticalFlip(p=0.5),
        # Image rotations
        RandomRotation(degrees=20),
        # Image shear and shift of up to 20%
        RandomAffine(degrees=0, translate=(0.2, 0.2), shear=np.degrees(np.arctan(0.2))),
        # Image zoom of up to 20%
        RandomResizedCrop(size=IMAGE_SIZE, scale=(0.64, 1.0), ratio=(ASPECT, ASPECT)),
    ]
)

suspicion cutpoint pos_weights:  [2.0, 9.53, 19.22, 20.0]
findings class weights:  [0.27, 1.51, 0.8, 1.36, 0.4, 0.53, 0.88, 1.13, 1.11, 2.02]
density class weights:  [3.5, 1.2, 0.94, 1.08]

focal alpha 0.0495 -> malignant weight 0.0495, benign weight 0.9505 (pos:neg = 0.0520)


In [6]:
cnn_train_dataset = MammoCNNDataset(cnn_df, transform=augmentations, is_training=True)
val_dataset = MammoCNNDataset(val_df, is_training=False)
test_dataset = MammoCNNDataset(test_df, is_training=False)

cnn_train_loader = DataLoader(
    cnn_train_dataset,
    batch_size=BATCH_SIZE,
    shuffle=True,
    num_workers=NUM_WORKERS,
    pin_memory=True,
    worker_init_fn=seed_worker,
    generator=g,
    prefetch_factor=2,
    persistent_workers=True,
)
val_loader = DataLoader(
    val_dataset,
    batch_size=BATCH_SIZE,
    shuffle=False,
    num_workers=NUM_WORKERS,
    pin_memory=True,
    worker_init_fn=seed_worker,
    generator=g,
    prefetch_factor=2,
    persistent_workers=True
)
test_loader = DataLoader(
    test_dataset,
    batch_size=BATCH_SIZE,
    shuffle=False,
    num_workers=NUM_WORKERS,
    pin_memory=True,
    worker_init_fn=seed_worker,
    generator=g,
    prefetch_factor=2,
    persistent_workers=True
)

# Examine the train dataset
patient_batch = next(iter(cnn_train_loader))

print(f"Image Shape:     {patient_batch['mammogram'].shape}")
print(f"Diagnosis Shape: {patient_batch['diagnosis'].shape} {patient_batch['diagnosis'].dtype}")
print(f"Findings Shape:  {patient_batch['findings'].shape} {patient_batch['findings'].dtype}")
print(f"Suspicion Shape: {patient_batch['suspicion'].shape} {patient_batch['suspicion'].dtype}")
print(f"Density Shape:   {patient_batch['density'].shape} {patient_batch['density'].dtype}")
print(f"Age Shape:       {patient_batch['age'].shape} {patient_batch['age'].dtype}")

# Gradient accumulation exists to push this probability down: a batch with no
# positive contributes nothing to the diagnosis gradient. Below ~0.05 the
# effective batch is large enough and ACCUMULATION_STEPS can go to 1.
positive_rate = cnn_df["diagnosis"].mean()
for n in (BATCH_SIZE, BATCH_SIZE * 2):
    print(
        f"\npositive rate {100 * positive_rate:.2f}%  ->  "
        f"P(no positive in a batch of {n}) = {(1 - positive_rate) ** n:.4f}"
    )

Image Shape:     torch.Size([48, 3, 1280, 1024])
Diagnosis Shape: torch.Size([48]) torch.float32
Findings Shape:  torch.Size([48, 10]) torch.float32
Suspicion Shape: torch.Size([48]) torch.int64
Density Shape:   torch.Size([48]) torch.int64
Age Shape:       torch.Size([48]) torch.float32

positive rate 4.95%  ->  P(no positive in a batch of 48) = 0.0877

positive rate 4.95%  ->  P(no positive in a batch of 96) = 0.0077


#### Train the MAMMO CNN (Phase 1)

**Loss scale calibration.** The paper states that "loss weighting was adjusted according to the auxiliary output losses, such that the loss weight for diagnosis was greater than or equal to the sum of all other auxiliary output losses" (Appendix C-B). `LOSS_WEIGHTS` encodes that intent, but a coefficient only equals a share of the gradient if the raw losses share a scale — and they do not. Focal loss shrinks toward zero by design; CORAL sums weighted BCE over 4 cutpoints. Uncorrected, `diagnosis: 0.50` lands at ~4% of the signal while suspicion takes ~70%.

`calibrate_loss_scales()` measures each task's magnitude once on the freshly-initialised model and freezes the reciprocals. The coefficients then mean what they say: diagnosis really is 50% at step 0. The constants are deliberately **not** a running average — re-normalising every step would keep amplifying a task that has already converged, so no auxiliary would ever be allowed to finish.

Watch the `loss share` line: it should start at roughly 50/15/15/10/10 and then drift as tasks converge at different rates. Drift is expected and correct; a task getting easier legitimately gives up share.

**Gradient accumulation** gives an effective batch of `BATCH_SIZE * ACCUMULATION_STEPS`. At 4.95% positives a batch of 32 has a 19.6% chance of containing no positive at all; 64 drops that to 3.8%. On an A100 you can raise `BATCH_SIZE` instead and set `ACCUMULATION_STEPS = 1`.

In [7]:
EPOCHS: int = 100
LEARNING_RATE: float = 3e-4  # head LR
BACKBONE_LR_MULT: float = 0.1  # backbone trains at LEARNING_RATE * this
WARMUP_EPOCHS: float = 2.0
WEIGHT_DECAY: float = 1e-4
ACCUMULATION_STEPS: int = 2  # effective batch = BATCH_SIZE * ACCUMULATION_STEPS = 96

mammo_cnn = MammoCNN(
    n_findings=findings_matrix.shape[1],
    n_suspicion=5,
    n_density=4,
    pretrained=True,
)

# Deviation from the paper: the whole network trains from step 1 rather than
# cycling frozen/unfrozen stages. The paper's stated reason for cycling was
# preserving pretrained ImageNet features (Appendix C-B, after Shen 2017), not
# a memory limit — so the underlying risk still applies here. WARMUP_EPOCHS and
# BACKBONE_LR_MULT address it instead: nothing runs at full LR until the random
# heads have settled, and the backbone moves at a tenth of the head rate
# throughout. Call mammo_cnn.freeze_backbone() to reproduce the paper's regime.
print("full network:", mammo_cnn.get_parameter_counts())

# ── Calibrate loss scales so LOSS_WEIGHTS means gradient share ────────────────
# Run once, on the freshly-initialised model, then freeze.
loss_scales = calibrate_loss_scales(
    mammo_cnn, cnn_train_loader, DEVICE, loss_kwargs, n_batches=10
)
loss_kwargs["loss_scales"] = loss_scales

print("\nloss scales (1 / measured magnitude):")
for task, scale in loss_scales.items():
    print(f"  {task:>10}: {scale:9.4f}   (raw magnitude {1 / scale:.4f})")

# Sanity check: with the scales applied, shares should match LOSS_WEIGHTS
check = effective_contributions(
    {task: 1 / scale for task, scale in loss_scales.items()}, LOSS_WEIGHTS, loss_scales
)
print("\nshare at step 0 vs intended:")
for task in loss_scales:
    print(f"  {task:>10}: {100 * check[task]:5.1f}%  (target {100 * LOSS_WEIGHTS[task]:.0f}%)")

full network: {'total': 55897780, 'trainable': 55897780}

loss scales (1 / measured magnitude):
   diagnosis:    4.8857   (raw magnitude 0.2047)
    findings:    5.5075   (raw magnitude 0.1816)
   suspicion:    0.1951   (raw magnitude 5.1248)
     density:    0.7363   (raw magnitude 1.3582)
         age:    1.0953   (raw magnitude 0.9130)

share at step 0 vs intended:
   diagnosis:  50.0%  (target 50%)
    findings:  15.0%  (target 15%)
   suspicion:  15.0%  (target 15%)
     density:  10.0%  (target 10%)
         age:  10.0%  (target 10%)


In [8]:
ALPHA_STRATEGY = "positive_rate" if ALPHA < 0.5 else "one_minus_positive_rate"

# Name and config both derive from the actual values — a run labelled with a
# resolution or an alpha it was not trained at is worse than no label.
run = wandb.init(
    project="tulip-mammo-cnn",
    name=f"phase1-{IMAGE_SIZE[0]}x{IMAGE_SIZE[1]}-alpha-{ALPHA_STRATEGY}-lr-{LEARNING_RATE}",
    config={
        "phase": "cnn-phase1",
        "backbone": "inception_resnet_v2",
        "pretrained": True,
        "image_size": IMAGE_SIZE,
        "batch_size": BATCH_SIZE,
        "accumulation_steps": ACCUMULATION_STEPS,
        "effective_batch": BATCH_SIZE * ACCUMULATION_STEPS,
        "epochs": EPOCHS,
        "learning_rate": LEARNING_RATE,
        "backbone_lr_mult": BACKBONE_LR_MULT,
        "backbone_lr": LEARNING_RATE * BACKBONE_LR_MULT,
        "warmup_epochs": WARMUP_EPOCHS,
        "weight_decay": WEIGHT_DECAY,
        "freeze_schedule": "none (full network from step 1)",
        "seed": SEED,
        "loss_weights": LOSS_WEIGHTS,
        "loss_scales": loss_scales,
        "alpha": ALPHA,
        "alpha_strategy": ALPHA_STRATEGY,
        "malignant_weight": ALPHA,
        "benign_weight": 1 - ALPHA,
        "gamma": GAMMA,
        "suspicion_head": "coral-ordinal",
        "sampler": "none (class weighting in loss)",
        "suspicion_cutpoint_weights": suspicion_weights.tolist(),
        "findings_class_weights": findings_weights.tolist(),
        "density_class_weights": density_weights.tolist(),
        "n_train_images": len(cnn_df),
        "n_val_images": len(val_df),
        "positive_rate": POSITIVE_RATE,
    },
)

# Gradient and parameter histograms. Worth it here specifically to watch the
# CORAL cutpoint biases: if they stop being well separated the ordinal head has
# collapsed to a single threshold and suspicion is no longer ordinal.
wandb.watch(mammo_cnn, log="all", log_freq=100)

print(f"logging to {run.url}")

wandb: [wandb.login()] Loaded credentials for https://api.wandb.ai from WANDB_API_KEY.
wandb: Currently logged in as: moyinoluwa (moyinoluwa-university-of-lagos) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


logging to https://wandb.ai/moyinoluwa-university-of-lagos/tulip-mammo-cnn/runs/nlrwqjhw


In [9]:
history = train(
    mammo_cnn,
    cnn_train_loader,
    val_loader,
    device=DEVICE,
    epochs=EPOCHS,
    learning_rate=LEARNING_RATE,
    weight_decay=WEIGHT_DECAY,
    accumulation_steps=ACCUMULATION_STEPS,
    warmup_epochs=WARMUP_EPOCHS,
    backbone_lr_mult=BACKBONE_LR_MULT,
    loss_kwargs=loss_kwargs,
    checkpoint_path=os.path.join("models", "mammo_cnn_phase1.pt"),
    monitor="ordinal_ap",
    monitor_mode="max",
    patience=7,
    lr_patience=3,
    lr_decay_factor=0.1,
    min_lr=1e-6,
    findings_names=CLINICAL_FINDINGS,
    age_std=AGE_STD,
    wandb_run=run,
)

# Store the best checkpoint as a versioned artifact so a run's weights stay
# tied to the config that produced them.
artifact = wandb.Artifact("mammo-cnn-phase1", type="model")
artifact.add_file(os.path.join("models", "mammo_cnn_phase1.pt"))
run.log_artifact(artifact)

epoch 1/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   1  train 0.4472  val 0.3669  lr[heads] 1.50e-04  lr[backbone] 1.50e-05  (warmup)
    diagnosis  AUC 0.4694  AP 0.0450   |  ordinal  AUC 0.5371  AP 0.0555
    findings   macroAP 0.0133  microAP 0.0467  macroAUC 0.4586
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0581  0.4966   105     0.97
      Suspicious Calcification      0.0233  0.4868    40     1.02
      Focal Asymmetry               0.0145  0.4350    28     0.91
      Architectural Distortion      0.0081  0.4845    13     1.09
      Asymmetry                     0.0028  0.3390     6     0.83
      Suspicious Lymph Node         0.0021  0.2399     5     0.75
      Nipple Retraction             0.0027  0.4526     4     1.18
      Skin Thickening               0.0063  0.7397     3     3.72
      Skin Retraction               0.0022  0.4532     3     1.26
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 0.84

epoch 2/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   2  train 0.3030  val 0.2852  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.4550  AP 0.0420   |  ordinal  AUC 0.6831  AP 0.1004
    findings   macroAP 0.0149  microAP 0.0446  macroAUC 0.5517
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0545  0.4758   105     0.91
      Suspicious Calcification      0.0237  0.5273    40     1.04
      Focal Asymmetry               0.0209  0.5533    28     1.32
      Architectural Distortion      0.0109  0.5135    13     1.48
      Asymmetry                     0.0118  0.6787     6     3.46
      Suspicious Lymph Node         0.0036  0.5534     5     1.26
      Nipple Retraction             0.0034  0.5831     4     1.51
      Skin Thickening               0.0027  0.5515     3     1.59
      Skin Retraction               0.0026  0.5289     3     1.55
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.647  exact 0

epoch 3/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   3  train 0.2693  val 0.2841  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.4079  AP 0.0369   |  ordinal  AUC 0.6643  AP 0.1568
    findings   macroAP 0.0132  microAP 0.0434  macroAUC 0.4998
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0557  0.4965   105     0.93
      Suspicious Calcification      0.0191  0.4362    40     0.84
      Focal Asymmetry               0.0194  0.5442    28     1.22
      Architectural Distortion      0.0086  0.5437    13     1.16
      Asymmetry                     0.0059  0.6380     6     1.74
      Suspicious Lymph Node         0.0025  0.3519     5     0.87
      Nipple Retraction             0.0026  0.4896     4     1.16
      Skin Thickening               0.0031  0.5566     3     1.79
      Skin Retraction               0.0019  0.4413     3     1.14
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.942  exact 0

epoch 4/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   4  train 0.2525  val 0.6559  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6471  AP 0.0966   |  ordinal  AUC 0.6581  AP 0.1172
    findings   macroAP 0.0275  microAP 0.0771  macroAUC 0.5842
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1189  0.5972   105     1.99
      Suspicious Calcification      0.0573  0.6005    40     2.52
      Focal Asymmetry               0.0191  0.5491    28     1.20
      Architectural Distortion      0.0078  0.5021    13     1.06
      Asymmetry                     0.0071  0.6117     6     2.08
      Suspicious Lymph Node         0.0036  0.5397     5     1.26
      Nipple Retraction             0.0217  0.6140     4     9.55
      Skin Thickening               0.0038  0.5064     3     2.22
      Skin Retraction               0.0083  0.7369     3     4.87
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 3.528  exact 0

epoch 5/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   5  train 0.2412  val 0.2729  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.4782  AP 0.1497   |  ordinal  AUC 0.6792  AP 0.2454
    findings   macroAP 0.0487  microAP 0.1154  macroAUC 0.5797
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1395  0.5532   105     2.34
      Suspicious Calcification      0.1986  0.6493    40     8.74
      Focal Asymmetry               0.0222  0.5648    28     1.39
      Architectural Distortion      0.0082  0.5409    13     1.12
      Asymmetry                     0.0043  0.5479     6     1.27
      Suspicious Lymph Node         0.0048  0.4572     5     1.70
      Nipple Retraction             0.0240  0.7298     4    10.57
      Skin Thickening               0.0020  0.4297     3     1.16
      Skin Retraction               0.0347  0.7445     3    20.38
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.770  exact 0

epoch 6/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   6  train 0.2338  val 0.2678  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.5490  AP 0.1948   |  ordinal  AUC 0.6648  AP 0.2260
    findings   macroAP 0.0469  microAP 0.1211  macroAUC 0.5743
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1384  0.5969   105     2.32
      Suspicious Calcification      0.2134  0.6342    40     9.39
      Focal Asymmetry               0.0235  0.6068    28     1.48
      Architectural Distortion      0.0078  0.5092    13     1.05
      Asymmetry                     0.0033  0.4359     6     0.97
      Suspicious Lymph Node         0.0107  0.4431     5     3.78
      Nipple Retraction             0.0147  0.7400     4     6.47
      Skin Thickening               0.0026  0.5147     3     1.54
      Skin Retraction               0.0074  0.6883     3     4.31
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.702  exact 0

epoch 7/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   7  train 0.2261  val 0.3103  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7076  AP 0.3422   |  ordinal  AUC 0.8077  AP 0.3240
    findings   macroAP 0.0749  microAP 0.1793  macroAUC 0.7195
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2411  0.7244   105     4.04
      Suspicious Calcification      0.2771  0.8038    40    12.19
      Focal Asymmetry               0.0512  0.7024    28     3.22
      Architectural Distortion      0.0120  0.5159    13     1.63
      Asymmetry                     0.0085  0.7299     6     2.51
      Suspicious Lymph Node         0.0300  0.8542     5    10.57
      Nipple Retraction             0.0309  0.8954     4    13.58
      Skin Thickening               0.0042  0.4278     3     2.46
      Skin Retraction               0.0189  0.8215     3    11.07
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.657  exact 0

epoch 8/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   8  train 0.2181  val 0.2796  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6388  AP 0.3018   |  ordinal  AUC 0.7591  AP 0.3686
    findings   macroAP 0.0683  microAP 0.1802  macroAUC 0.6634
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2387  0.6847   105     4.00
      Suspicious Calcification      0.2716  0.7442    40    11.95
      Focal Asymmetry               0.0346  0.6825    28     2.18
      Architectural Distortion      0.0102  0.5491    13     1.38
      Asymmetry                     0.0051  0.6143     6     1.51
      Suspicious Lymph Node         0.0218  0.5938     5     7.68
      Nipple Retraction             0.0152  0.7524     4     6.68
      Skin Thickening               0.0081  0.5690     3     4.76
      Skin Retraction               0.0095  0.7805     3     5.57
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.456  exact 0

epoch 9/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   9  train 0.2152  val 0.2886  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7611  AP 0.4107   |  ordinal  AUC 0.7943  AP 0.4155
    findings   macroAP 0.0894  microAP 0.2035  macroAUC 0.7679
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2869  0.7429   105     4.81
      Suspicious Calcification      0.3204  0.7858    40    14.10
      Focal Asymmetry               0.0624  0.7346    28     3.92
      Architectural Distortion      0.0152  0.5914    13     2.06
      Asymmetry                     0.0078  0.6463     6     2.29
      Suspicious Lymph Node         0.0331  0.9053     5    11.65
      Nipple Retraction             0.0404  0.8781     4    17.76
      Skin Thickening               0.0079  0.6911     3     4.66
      Skin Retraction               0.0304  0.9357     3    17.83
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.599  exact 0

epoch 10/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  10  train 0.2086  val 0.3250  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7917  AP 0.3671   |  ordinal  AUC 0.8113  AP 0.3625
    findings   macroAP 0.0825  microAP 0.1901  macroAUC 0.7440
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2652  0.7367   105     4.45
      Suspicious Calcification      0.3186  0.8116    40    14.02
      Focal Asymmetry               0.0618  0.6816    28     3.89
      Architectural Distortion      0.0210  0.6135    13     2.84
      Asymmetry                     0.0101  0.6084     6     2.96
      Suspicious Lymph Node         0.0310  0.8538     5    10.91
      Nipple Retraction             0.0147  0.8253     4     6.47
      Skin Thickening               0.0052  0.6568     3     3.07
      Skin Retraction               0.0145  0.9080     3     8.48
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.484  exact 0

epoch 11/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  11  train 0.2073  val 0.3206  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7524  AP 0.3206   |  ordinal  AUC 0.8065  AP 0.3330
    findings   macroAP 0.0682  microAP 0.1703  macroAUC 0.7078
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2723  0.7184   105     4.56
      Suspicious Calcification      0.2150  0.7764    40     9.46
      Focal Asymmetry               0.0564  0.7373    28     3.55
      Architectural Distortion      0.0173  0.5883    13     2.35
      Asymmetry                     0.0059  0.6233     6     1.72
      Suspicious Lymph Node         0.0184  0.7150     5     6.48
      Nipple Retraction             0.0150  0.7788     4     6.61
      Skin Thickening               0.0035  0.6261     3     2.07
      Skin Retraction               0.0097  0.8071     3     5.71
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.961  exact 0

epoch 12/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  12  train 0.2046  val 0.3295  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7884  AP 0.3207   |  ordinal  AUC 0.7962  AP 0.3399
    findings   macroAP 0.0697  microAP 0.1645  macroAUC 0.7018
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2607  0.7174   105     4.37
      Suspicious Calcification      0.2453  0.7987    40    10.80
      Focal Asymmetry               0.0522  0.6779    28     3.28
      Architectural Distortion      0.0161  0.6052    13     2.18
      Asymmetry                     0.0060  0.5697     6     1.77
      Suspicious Lymph Node         0.0185  0.7828     5     6.53
      Nipple Retraction             0.0155  0.7624     4     6.82
      Skin Thickening               0.0044  0.6352     3     2.59
      Skin Retraction               0.0089  0.7672     3     5.21
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.711  exact 0

epoch 13/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  13  train 0.2030  val 0.3257  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7481  AP 0.3721   |  ordinal  AUC 0.8013  AP 0.3896
    findings   macroAP 0.0754  microAP 0.1595  macroAUC 0.7580
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2887  0.7437   105     4.84
      Suspicious Calcification      0.2402  0.7566    40    10.57
      Focal Asymmetry               0.0577  0.7479    28     3.62
      Architectural Distortion      0.0128  0.5595    13     1.73
      Asymmetry                     0.0115  0.6909     6     3.37
      Suspicious Lymph Node         0.0295  0.8145     5    10.37
      Nipple Retraction             0.0132  0.7835     4     5.79
      Skin Thickening               0.0077  0.7923     3     4.53
      Skin Retraction               0.0175  0.9330     3    10.26
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 3.086  exact 0

epoch 14/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  14  train 0.1962  val 0.3121  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7834  AP 0.4062   |  ordinal  AUC 0.8183  AP 0.4338
    findings   macroAP 0.0835  microAP 0.1881  macroAUC 0.7905
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3093  0.7676   105     5.18
      Suspicious Calcification      0.2740  0.7859    40    12.06
      Focal Asymmetry               0.0558  0.7442    28     3.51
      Architectural Distortion      0.0129  0.5947    13     1.74
      Asymmetry                     0.0145  0.7851     6     4.26
      Suspicious Lymph Node         0.0409  0.8712     5    14.41
      Nipple Retraction             0.0182  0.8410     4     8.01
      Skin Thickening               0.0078  0.7856     3     4.57
      Skin Retraction               0.0184  0.9395     3    10.81
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.895  exact 0

epoch 15/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  15  train 0.1903  val 0.2788  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7884  AP 0.3899   |  ordinal  AUC 0.8243  AP 0.4200
    findings   macroAP 0.0768  microAP 0.1811  macroAUC 0.7668
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2835  0.7509   105     4.75
      Suspicious Calcification      0.2509  0.7914    40    11.04
      Focal Asymmetry               0.0509  0.7312    28     3.20
      Architectural Distortion      0.0124  0.5552    13     1.67
      Asymmetry                     0.0108  0.7287     6     3.18
      Suspicious Lymph Node         0.0417  0.8288     5    14.67
      Nipple Retraction             0.0139  0.7920     4     6.13
      Skin Thickening               0.0063  0.7748     3     3.71
      Skin Retraction               0.0210  0.9486     3    12.33
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.479  exact 0

epoch 16/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  16  train 0.1874  val 0.2785  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7917  AP 0.3978   |  ordinal  AUC 0.8169  AP 0.4145
    findings   macroAP 0.0796  microAP 0.1875  macroAUC 0.7659
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2929  0.7493   105     4.91
      Suspicious Calcification      0.2583  0.7881    40    11.37
      Focal Asymmetry               0.0544  0.7383    28     3.42
      Architectural Distortion      0.0125  0.5582    13     1.70
      Asymmetry                     0.0121  0.7298     6     3.54
      Suspicious Lymph Node         0.0448  0.8464     5    15.77
      Nipple Retraction             0.0152  0.8106     4     6.70
      Skin Thickening               0.0057  0.7245     3     3.36
      Skin Retraction               0.0207  0.9482     3    12.14
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.402  exact 0

epoch 17/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  17  train 0.1881  val 0.2966  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7817  AP 0.3943   |  ordinal  AUC 0.8106  AP 0.4080
    findings   macroAP 0.0781  microAP 0.1850  macroAUC 0.7529
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2777  0.7444   105     4.66
      Suspicious Calcification      0.2692  0.7929    40    11.84
      Focal Asymmetry               0.0561  0.7488    28     3.53
      Architectural Distortion      0.0136  0.5562    13     1.84
      Asymmetry                     0.0112  0.7207     6     3.28
      Suspicious Lymph Node         0.0399  0.8083     5    14.05
      Nipple Retraction             0.0143  0.7778     4     6.28
      Skin Thickening               0.0050  0.6995     3     2.92
      Skin Retraction               0.0159  0.9277     3     9.35
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.603  exact 0

epoch 18/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  18  train 0.1867  val 0.2680  lr[heads] 3.00e-06  lr[backbone] 1.00e-06
    diagnosis  AUC 0.7871  AP 0.3887   |  ordinal  AUC 0.8183  AP 0.4214
    findings   macroAP 0.0771  microAP 0.1825  macroAUC 0.7594
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2925  0.7438   105     4.90
      Suspicious Calcification      0.2516  0.7975    40    11.07
      Focal Asymmetry               0.0559  0.7300    28     3.51
      Architectural Distortion      0.0135  0.5557    13     1.82
      Asymmetry                     0.0085  0.7019     6     2.50
      Suspicious Lymph Node         0.0381  0.8520     5    13.42
      Nipple Retraction             0.0155  0.8218     4     6.83
      Skin Thickening               0.0049  0.7228     3     2.86
      Skin Retraction               0.0137  0.9089     3     8.01
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.205  exact 0

epoch 19/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  19  train 0.1856  val 0.2869  lr[heads] 3.00e-06  lr[backbone] 1.00e-06
    diagnosis  AUC 0.7876  AP 0.4076   |  ordinal  AUC 0.8084  AP 0.4252
    findings   macroAP 0.0805  microAP 0.1895  macroAUC 0.7776
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3044  0.7531   105     5.10
      Suspicious Calcification      0.2666  0.7980    40    11.73
      Focal Asymmetry               0.0513  0.7328    28     3.22
      Architectural Distortion      0.0127  0.5673    13     1.73
      Asymmetry                     0.0091  0.7122     6     2.68
      Suspicious Lymph Node         0.0380  0.8772     5    13.38
      Nipple Retraction             0.0209  0.8592     4     9.20
      Skin Thickening               0.0058  0.7826     3     3.41
      Skin Retraction               0.0152  0.9158     3     8.90
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.525  exact 0

epoch 20/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  20  train 0.1864  val 0.2950  lr[heads] 3.00e-06  lr[backbone] 1.00e-06
    diagnosis  AUC 0.7928  AP 0.3948   |  ordinal  AUC 0.8165  AP 0.4171
    findings   macroAP 0.0789  microAP 0.1836  macroAUC 0.7604
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2988  0.7507   105     5.01
      Suspicious Calcification      0.2566  0.7968    40    11.29
      Focal Asymmetry               0.0565  0.7423    28     3.55
      Architectural Distortion      0.0139  0.5699    13     1.89
      Asymmetry                     0.0114  0.6944     6     3.33
      Suspicious Lymph Node         0.0378  0.8604     5    13.31
      Nipple Retraction             0.0171  0.8293     4     7.54
      Skin Thickening               0.0046  0.6830     3     2.73
      Skin Retraction               0.0131  0.9163     3     7.67
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.591  exact 0

epoch 21/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  21  train 0.1843  val 0.2877  lr[heads] 3.00e-06  lr[backbone] 1.00e-06
    diagnosis  AUC 0.7933  AP 0.3986   |  ordinal  AUC 0.8138  AP 0.4204
    findings   macroAP 0.0790  microAP 0.1867  macroAUC 0.7713
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2982  0.7518   105     5.00
      Suspicious Calcification      0.2567  0.7999    40    11.30
      Focal Asymmetry               0.0584  0.7354    28     3.67
      Architectural Distortion      0.0136  0.5754    13     1.85
      Asymmetry                     0.0097  0.7046     6     2.86
      Suspicious Lymph Node         0.0379  0.8656     5    13.35
      Nipple Retraction             0.0172  0.8410     4     7.55
      Skin Thickening               0.0056  0.7479     3     3.27
      Skin Retraction               0.0140  0.9203     3     8.19
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.522  exact 0

<Artifact mammo-cnn-phase1>

In [10]:
# Per-class findings report
final = evaluate(
    mammo_cnn, val_loader, DEVICE, loss_kwargs, CLINICAL_FINDINGS, AGE_STD
)

report = pd.DataFrame(
    {
        "support": final["findings_support"],
        "AUROC": final["findings_auc_per_class"],
        "AP": final["findings_ap_per_class"],
    }
).sort_values("support", ascending=False)
report["prevalence %"] = 100 * report["support"] / len(val_df)

print("── findings, per class (validation) ──")
print(report.to_string(float_format=lambda v: f"{v:.4f}"))
print(f"\nmacro AP {final['findings_macro_ap']:.4f}   micro AP {final['findings_micro_ap']:.4f}")

print("\n── suspicion, per BI-RADS level (paper Table I format) ──")
for level, auc in final["suspicion_auc_per_level"].items():
    print(f"  {level}: AUROC {auc:.4f}")

print("\n── comparison with the paper (Tommy dataset, Table I) ──")
print(f"  diagnosis AUROC   ours {final['ordinal_auc']:.3f}   paper 0.787")
print(f"  diagnosis AUPRC   ours {final['ordinal_ap']:.3f}   paper 0.394")
print(f"  age MAE (years)   ours {final['age_mae_years']:.2f}    paper 5.97")
print("  (the paper used TTA over 100 samples; these are single-pass)")

run.finish()

── findings, per class (validation) ──
                          support  AUROC     AP  prevalence %
Mass                          105 0.7676 0.3093        5.9659
Suspicious Calcification       40 0.7859 0.2740        2.2727
Focal Asymmetry                28 0.7442 0.0558        1.5909
Architectural Distortion       13 0.5947 0.0129        0.7386
Asymmetry                       6 0.7851 0.0145        0.3409
Suspicious Lymph Node           5 0.8712 0.0409        0.2841
Nipple Retraction               4 0.8410 0.0182        0.2273
Skin Thickening                 3 0.7856 0.0078        0.1705
Skin Retraction                 3 0.9395 0.0184        0.1705
Global Asymmetry                0    NaN    NaN        0.0000

macro AP 0.0835   micro AP 0.1881

── suspicion, per BI-RADS level (paper Table I format) ──
  birads_1: AUROC 0.6437
  birads_2: AUROC 0.4530
  birads_3: AUROC 0.3275
  birads_4: AUROC 0.2220
  birads_5: AUROC 0.9604

── comparison with the paper (Tommy dataset, Table I) ──
  

epoch,▁▁▂▂▂▃▃▃▄▄▅▅▅▆▆▆▇▇▇██
train/loss,█▄▃▃▃▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁
train/loss_share/age,▁█▆▆▇▇▇▇▇▇▇▆▆▆▅▆▆▆▆▅▆
train/loss_share/density,▂█▂▁▁▁▁▁▂▁▁▁▁▃▃▂▂▂▃▂▃
train/loss_share/diagnosis,█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
train/loss_share/findings,█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
train/loss_share/suspicion,▁▆███████████████████
train/lr/backbone,▄███████████▁▁▁▁▁▁▁▁▁
train/lr/heads,▄███████████▂▂▂▂▂▁▁▁▁
train/raw_loss/age,█▆▄▄▃▃▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁
+54,...
